File: src/src_core/fmg/core/mosaic/MosaicGameModel.cs
////////////////////////////////////////////////////////////////////////////////
//                               FastMines project
// file name: "MosaicGameModel.cs"
//
// Author: 2011-2018  -  Serhii Kryvulia aka SeregaLBN
//
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License
// as published by the Free Software Foundation; either version 2
// of the License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
////////////////////////////////////////////////////////////////////////////////
using System;
using System.Linq;
using System.Collections.Generic;
using System.ComponentModel;
using fmg.common;
using fmg.common.geom;
using fmg.common.notyfier;
using fmg.core.types;
using fmg.core.mosaic.cells;

namespace fmg.core.mosaic {

    /// <summary> MVC: game model of mosaic field. Default implementation </summary>
    public class MosaicGameModel : IMosaic, INotifyPropertyChanged, IDisposable {

    #region Members

        private BaseCell.BaseAttribute _cellAttr;
        /// <summary> Matrix of cells, is represented as a vector <see cref="IList<BaseCell>"/>.
        /// Матрица ячеек, представленная(развёрнута) в виде вектора </summary>
        private readonly IList<BaseCell> _matrix = new List<BaseCell>();
        /// <summary> Field size in cells </summary>
        private Matrisize _sizeField = new Matrisize(10, 10);
        /// <summary> из каких фигур состоит мозаика поля </summary>
        private EMosaic _mosaicType = EMosaic.eMosaicSquare1;

        protected bool Disposed { get; private set; }
        private event PropertyChangedEventHandler PropertyChangedSync;
        public  event PropertyChangedEventHandler PropertyChanged/*Async*/;
        protected readonly NotifyPropertyChanged _notifier;
        private   readonly NotifyPropertyChanged _notifierAsync;
    #endregion

        public MosaicGameModel() {
            _notifier      = new NotifyPropertyChanged(this, ev => PropertyChangedSync?.Invoke(this, ev), false);
            _notifierAsync = new NotifyPropertyChanged(this, ev => PropertyChanged    ?.Invoke(this, ev), true);
            this.PropertyChangedSync += OnPropertyChanged;
        }

        public BaseCell.BaseAttribute CellAttr {
            get {
                if (_cellAttr == null) {
                    _cellAttr = MosaicHelper.CreateAttributeInstance(MosaicType);
                    _cellAttr.PropertyChanged += OnCellAttributePropertyChanged;
                }
                return _cellAttr;
            }
            set {
                if (_cellAttr == null)
                    return;
                if (value != null)
                    throw new ArgumentException("Bad argument - support only null value!");
                _cellAttr.PropertyChanged -= OnCellAttributePropertyChanged;
                _cellAttr = null;
                _matrix.Clear();
                _notifier.FirePropertyChanged();
                _notifier.FirePropertyChanged(nameof(this.Matrix));
            }
        }

        /// <summary> площадь ячеек </summary>
        public double Area {
            get { return CellAttr.Area; }
            set {
                if (value <= 0)
                    throw new ArgumentException("Area must be positive");
                CellAttr.Area = value;
            }
        }

        public IList<BaseCell> Matrix {
            get {
                if (!_matrix.Any()) {
                    var attr = CellAttr;
                    var size = SizeField;
                    var mosaicType = MosaicType;
                    //_matrix = new List<BaseCell>(size.width * size.height);
                    for (var i=0; i < size.m; i++)
                        for (var j=0; j < size.n; j++) {
                            var cell = MosaicHelper.CreateCellInstance(attr, mosaicType, new Coord(i, j));
                            _matrix.Add(/* i*_size.n + j, */ cell);
                        }
                }
                return _matrix;
            }
        }

        /// <summary> mosaic field size in cells </summary>
        public Matrisize SizeField {
            get { return _sizeField; }
            set {
                var old = this._sizeField;
                if (old == value)
                    return;

                _matrix.Clear();
                this._sizeField = value;

                _notifier.FirePropertyChanged(old, value);
                _notifier.FirePropertyChanged(nameof(this.Matrix));
            }
        }

        /// <summary> mosaic type (из каких фигур состоит мозаика поля) </summary>
        public EMosaic MosaicType {
            get { return _mosaicType; }
            set {
                EMosaic old = this._mosaicType;
                if (old == value)
                    return;

                this._mosaicType = value;
                CellAttr = null;
                _notifier.FirePropertyChanged(old, value);
            }
        }

        /// <summary> доступ к заданной ячейке </summary>
        public BaseCell GetCell(int x, int y) { return Matrix[x*_sizeField.n + y]; }
        /// <summary> доступ к заданной ячейке </summary>
        public BaseCell GetCell(Coord coord) { return GetCell(coord.x, coord.y); }

        protected virtual void OnPropertyChanged(object sender, PropertyChangedEventArgs ev) {
            // refire as async event
            _notifierAsync.FirePropertyChanged(ev);
        }

        protected virtual void OnCellAttributePropertyChanged(object sender, PropertyChangedEventArgs ev) {
            System.Diagnostics.Debug.Assert(ReferenceEquals(sender, CellAttr));

            if (ev.PropertyName == nameof(BaseCell.BaseAttribute.Area)) {
                foreach (var cell in Matrix)
                    cell.Init();
                _notifier.FirePropertyChanged<double>(ev, nameof(this.Area)); // ! rethrow event - notify parent class
            }
            _notifier.FirePropertyChanged(nameof(this.CellAttr));
        }

        /// <summary> off notifier </summary>
        protected virtual IDisposable Hold() {
            var a1 = _notifier.Hold();
            var a2 = CellAttr.Hold();
            return new PlainFree() {
                _onDispose = () => {
                    a1.Dispose();
                    a2.Dispose();
                }
            };
        }

        /// <summary>  Dispose managed resources </summary>/
        protected virtual void Disposing() {
            this.PropertyChangedSync -= OnPropertyChanged;
            _notifier.Dispose();
            _notifierAsync.Dispose();
            CellAttr = null; // call setter - unsubscribe & dispose
        }

        public void Dispose() {
            if (Disposed)
                return;
            Disposed = true;
            Disposing();
            GC.SuppressFinalize(this);
        }

    }

}

File: FastMines_core/src/fmg/core/mosaic/MosaicGameModel.java
////////////////////////////////////////////////////////////////////////////////
//                               FastMines project
// file name: "MosaicGameModel.java"
//
// Data class: game model of mosaic field
// Author: 2011-2018  -  Serhii Kryvulia aka SeregaLBN
//
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License
// as published by the Free Software Foundation; either version 2
// of the License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
////////////////////////////////////////////////////////////////////////////////
package fmg.core.mosaic;

import java.beans.PropertyChangeEvent;
import java.beans.PropertyChangeListener;
import java.util.ArrayList;
import java.util.List;

import fmg.common.Logger;
import fmg.common.geom.Coord;
import fmg.common.geom.Matrisize;
import fmg.common.notifier.INotifyPropertyChanged;
import fmg.common.notifier.NotifyPropertyChanged;
import fmg.core.app.model.MosaicInitData;
import fmg.core.mosaic.cells.BaseCell;
import fmg.core.mosaic.shape.BaseShape;
import fmg.core.types.EMosaic;
import fmg.core.types.Property;

/** MVC: game model of mosaic field. Default implementation. */
@Deprecated
public class MosaicGameModel implements IMosaic, INotifyPropertyChanged, AutoCloseable {

    public static final String PROPERTY_AREA        = BaseShape.PROPERTY_AREA;
    public static final String PROPERTY_CELL_SHAPE  = "CellShape";
    public static final String PROPERTY_SIZE_FIELD  = "SizeField";
    public static final String PROPERTY_MATRIX      = "Matrix";
    public static final String PROPERTY_MOSAIC_TYPE = "MosaicType";

    @Property(PROPERTY_CELL_SHAPE)
    private BaseShape shape;

    /** Matrix of cells, is represented as a vector {@link List<BaseCell>}.
      * Матрица ячеек, представленная(развёрнута) в виде вектора */
    @Property(PROPERTY_MATRIX)
    private final List<BaseCell> matrix = new ArrayList<>();

    /** Field size in cells */
    @Property(PROPERTY_SIZE_FIELD)
    private Matrisize sizeField = new Matrisize(10, 10);

    /** из каких фигур состоит мозаика поля */
    @Property(PROPERTY_MOSAIC_TYPE)
    private EMosaic mosaicType = EMosaic.eMosaicSquare1;

    protected final NotifyPropertyChanged notifier/*Sync*/ = new NotifyPropertyChanged(this, false);
    private   final NotifyPropertyChanged notifierAsync    = new NotifyPropertyChanged(this, true);

    private final PropertyChangeListener      onPropertyChangedListener = this::onPropertyChanged;
    private final PropertyChangeListener onShapePropertyChangedListener = this::onShapePropertyChanged;

    public MosaicGameModel() {
        notifier.addListener(onPropertyChangedListener);
    }

    @Override
    public BaseShape getShape() {
        if (shape == null) {
            shape = MosaicHelper.createShapeInstance(getMosaicType());
            shape.addListener(onShapePropertyChangedListener);
        }
        return shape;
    }
    private void resetShape() {
        if (shape == null)
            return;
        shape.removeListener(onShapePropertyChangedListener);
        shape.close();
        shape = null;
        matrix.clear();
        notifier.firePropertyChanged(PROPERTY_CELL_SHAPE);
        notifier.firePropertyChanged(PROPERTY_MATRIX);
    }

    /** площадь ячеек */
    @Override
    public double getArea() {
        return getShape().getArea();
    }
    /** установить новую площадь ячеек */
    @Override
    public void setArea(double area) {
        if (area <= 0)
            throw new IllegalArgumentException("Area must be positive");
        if (area < MosaicInitData.AREA_MINIMUM)
            Logger.warn("The area is very small = " + area);
      //getShape().setArea(Math.max(MosaicInitData.AREA_MINIMUM, area));
        getShape().setArea(area);
    }

    @Override
    public List<BaseCell> getMatrix() {
        if (matrix.isEmpty()) {
            BaseShape s = getShape();
            Matrisize size = getSizeField();
            //_matrix = new ArrayList<BaseCell>(size.width * size.height);
            for (int i=0; i < size.m; i++)
                for (int j=0; j < size.n; j++) {
                    BaseCell cell = MosaicHelper.createCellInstance(s, mosaicType, new Coord(j, i));
                    matrix.add(/* i*_size.n + j, */ cell);
                }
        }
        return matrix;
    }

    /** get mosaic field size in cells */
    @Override
    public Matrisize getSizeField() { return sizeField; }
    /** set mosaic field size in cells */
    @Override
    public void setSizeField(Matrisize newSizeField) {
        Matrisize old = this.sizeField;
        if (old.equals(newSizeField))
            return;

        matrix.clear();
        this.sizeField = newSizeField;

        notifier.firePropertyChanged(old, newSizeField, PROPERTY_SIZE_FIELD);
        notifier.firePropertyChanged(PROPERTY_MATRIX);
    }

    /** get mosaic type
      * (из каких фигур состоит мозаика поля) */
    @Override
    public EMosaic getMosaicType() { return mosaicType; }
    /** set mosaic type */
    @Override
    public void setMosaicType(EMosaic newMosaicType) {
        EMosaic old = this.mosaicType;
        if (old == newMosaicType)
            return;

        this.mosaicType = newMosaicType;
        resetShape();
        notifier.firePropertyChanged(old, newMosaicType, PROPERTY_MOSAIC_TYPE);
    }

    /** доступ к заданной ячейке */
    public BaseCell getCell(int x, int y) { return getMatrix().get(x*sizeField.n + y); }
    /** доступ к заданной ячейке */
    @Override
    public BaseCell getCell(Coord coord) { return getCell(coord.x, coord.y); }

    protected void onPropertyChanged(PropertyChangeEvent ev) {
        // refire as async event
        notifierAsync.firePropertyChanged(ev.getOldValue(), ev.getNewValue(), ev.getPropertyName());
    }

    protected void onShapePropertyChanged(PropertyChangeEvent ev) {
        String propName = ev.getPropertyName();
        if (BaseShape.PROPERTY_AREA.equals(propName)) {
            getMatrix().forEach(BaseCell::init);

            notifier.firePropertyChanged(ev.getOldValue(), ev.getNewValue(), PROPERTY_AREA); // ! rethrow event - notify parent class
        }
        notifier.firePropertyChanged(PROPERTY_CELL_SHAPE);
    }

    @Override
    public void close() {
        notifier.removeListener(onPropertyChangedListener);
        notifier.close();
        notifierAsync.close();
        resetShape();
    }

    @Override
    public void addListener(PropertyChangeListener listener) {
        notifierAsync.addListener(listener);
    }
    @Override
    public void removeListener(PropertyChangeListener listener) {
        notifierAsync.removeListener(listener);
    }

}
